File: languages/cpp/04-expert/smart-pointers-in-depth/example/main.cpp
// Module focus: Tracking ownership and lifetime when multiple references can observe the same
// value. Why it matters: the example makes it possible to model exclusive, shared, and non-owning
// relationships idiomatically before the learner tackles the exercises.

// This example shows tracking ownership and lifetime when multiple references can observe the same
// value. In C++, the example keeps value flow, references, and explicit control visible.

#include <iostream>
#include <memory>
#include <string>
using namespace std;

class Document {
  public:
    explicit Document(const string& nameValue) : name(nameValue) {
        cout << "Created: " << name << '\n';
    }

    ~Document() { cout << "Destroyed: " << name << '\n'; }

  private:
    string name;
};

// Fixed inputs make the consequence of overusing `shared_ptr` where `unique_ptr` is sufficient
// visible and repeatable.
int main() {
    // These values exercise the normal path before the exercises vary the documented boundaries.
    unique_ptr<Document> owner(new Document("DesignDoc"));

    shared_ptr<Document> teamA(new Document("SharedSpec"));
    shared_ptr<Document> teamB = teamA;
    weak_ptr<Document> observer = teamA;

    // The printed result shows whether the program can prevent leaks, cycles, and stale
    // observations in ownership graphs.
    cout << "Shared use count: " << teamA.use_count() << '\n';

    if (shared_ptr<Document> locked = observer.lock()) {
        cout << "Observer can access shared document.\n";
        (void)locked;
    }

    owner.reset();
    teamA.reset();
    teamB.reset();

    if (observer.expired()) {
        cout << "Observer is expired after shared owners released.\n";
    }

    return 0;
}

File: languages/typescript/04-expert/smart-pointers-in-depth/example/main.ts
// Module focus: Tracking ownership and lifetime when multiple references can observe the same value.
// Why it matters: the example makes it possible to model exclusive, shared, and non-owning
// relationships idiomatically before the learner tackles the exercises.

// Separate helpers keep the main path focused on how to model exclusive, shared, and non-owning
// relationships idiomatically.
type Note = {
    id: string;
    body: string;
};

class NoteOwner {
    constructor(
        readonly label: string,
        private note: Note | null,
    ) {}

    transferTo(other: NoteOwner): void {
        // Moving ownership leaves the source empty and makes the destination responsible.
        if (this.note === null) {
            return;
        }

        other.note = this.note;
        this.note = null;
    }

    currentNote(): Note | null {
        // Returning the object shows that aliases can still observe shared mutable state.
        return this.note;
    }

    describe(): string {
        return this.note === null
            ? `${this.label} owns nothing`
            : `${this.label} owns ${this.note.id}`;
    }
}

// Fixed inputs make the consequence of treating object references as automatic deep copies
// visible and repeatable.
function main(): void {
    // These values exercise the normal path before the exercises vary the documented boundaries.
    const originalNote: Note = {
        id: "note-101",
        body: "Review release checklist",
    };

    const inbox = new NoteOwner("inbox", originalNote);
    const archive = new NoteOwner("archive", null);

    // Print ownership before and after transfer so the lifetime change is visible.
    // The printed result shows whether the program can prevent leaks, cycles, and stale
    // observations in ownership graphs.
    console.log(inbox.describe());
    console.log(archive.describe());

    inbox.transferTo(archive);

    console.log(inbox.describe());
    console.log(archive.describe());

    const alias = archive.currentNote();
    if (alias !== null) {
        alias.body = "Review release checklist and publish notes";
        console.log(`Alias sees: ${alias.body}`);
    }
}

main();

export {};
